### DISCUSSION TITLE
Help needed-Setting up HAFS DOCN workflow on Orion

### DISCUSSION THREAD

**Discussion by lliu1220** [hafs-community/HAFS#88] [Q&A] [answered]
I have submitted cronjob_hafs_docn.sh many times according to instruction.
Here is my rocoto, job stuck at run_forecast for hours then fail:
The path to my cronjob configuration:
/work/noaa/marine/lliu/HAFS_couplling/workflow/test/HAFS/rocoto
/work/noaa/marine/lliu/HAFS_couplling/workflow/test/HAFS/parm

Appreciate your attention and help.

       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
201908290000                  launch                     2763535           SUCCEEDED                   0         1           9.0
201908290000                   input                           -                   -                   -         -             -
201908290000        download_docn_00                     2763547           SUCCEEDED                   0         1         136.0
201908290000        download_docn_01                     2763548           SUCCEEDED                   0         1         136.0
201908290000                atm_prep                     2763549           SUCCEEDED                   0         1         635.0
201908290000                  atm_ic                     2770645           SUCCEEDED                   0         1         539.0
201908290000              atm_lbc001                     2770646           SUCCEEDED                   0         1         476.0
201908290000              atm_lbc002                     2770647           SUCCEEDED                   0         1         559.0
201908290000              atm_lbc003                     2770648           SUCCEEDED                   0         1         484.0
201908290000              atm_lbc004                     2770649           SUCCEEDED                   0         1         472.0
201908290000              atm_lbc005                     2770650           SUCCEEDED                   0         1         570.0
201908290000              atm_lbc006                     2770651           SUCCEEDED                   0         1         489.0
201908290000               docn_prep                     2770652           SUCCEEDED                   0         1         294.0
201908290000                forecast    druby://Orion-login-1.HPC.MsState.Edu:46522          SUBMITTING                   -         0           0.0
201908290000                atm_post                           -                   -                   -         -             -
201908290000                 product                           -                   -                   -         -             -
201908290000            archive_disk                           -                   -                   -         -             -

**Comment by lliu1220** [comment]
second attempt succeeded forecast it seems lilke it!

       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
201908290000                  launch                     2763535           SUCCEEDED                   0         1           9.0
201908290000                   input                           -                   -                   -         -             -
201908290000        download_docn_00                     2763547           SUCCEEDED                   0         1         136.0
201908290000        download_docn_01                     2763548           SUCCEEDED                   0         1         136.0
201908290000                atm_prep                     2763549           SUCCEEDED                   0         1         635.0
201908290000                  atm_ic                     2770645           SUCCEEDED                   0         1         539.0
201908290000              atm_lbc001                     2770646           SUCCEEDED                   0         1         476.0
201908290000              atm_lbc002                     2770647           SUCCEEDED                   0         1         559.0
201908290000              atm_lbc003                     2770648           SUCCEEDED                   0         1         484.0
201908290000              atm_lbc004                     2770649           SUCCEEDED                   0         1         472.0
201908290000              atm_lbc005                     2770650           SUCCEEDED                   0         1         570.0
201908290000              atm_lbc006                     2770651           SUCCEEDED                   0         1         489.0
201908290000               docn_prep                     2770652           SUCCEEDED                   0         1         294.0
201908290000                forecast                     2781463           SUCCEEDED                   0         1       14471.0
201908290000                atm_post                     2785280           SUCCEEDED                   0         1         468.0
201908290000                 product                     2785342           SUCCEEDED                   0         1         484.0
201908290000            archive_disk                     2785477              QUEUED                   -         0           0.0
201908290000            archive_tape                           -                   -                   -         -             -
201908290000          archive_fv3out                           -                   -                   -         -             -
201908290000              scrub_work                           -                   -                   -         -             -
201908290000               scrub_com                           -                   -                   -         -             -
201908290000                donefile                     2785478              QUEUED                   -         0           0.0
201908290000              completion                           -                   -                   -         -             -

**Reply by evankalina** [reply]
@lliu1220 Great to hear that it's working for you now. It is typical for the forecast to take a long time to run, because it requires a lot of cores for the atmosphere part. But it shouldn't fail. How did you get it to work?

**Reply by evankalina** [reply]
@lliu1220 Also, regarding your question:

"One last question, I promise--
I got 128 hours high resolution gfs_bndy.tile7.0*.nc:
        lon = 2888 ;
        lat = 1920 ;
        lonp = 2889 ;
        latm = 1919 ;

How to switch to lower resolution gfs?"

The resolution of the atmospheric lateral boundary conditions is controlled by chgres_cube. However, I think we can modify the workflow to obtain lower-resolution LBCs. What resolution are you hoping for?

**Reply by lliu1220** [reply]
@evankalina 
Thanks! I was hoping for this resolution:
netcdf gfs_bndy.tile7.024 {
dimensions:
        lon = 728 ;
        lat = 600 ;
        lonp = 729 ;
        latm = 599 ;
        halo = 14 ;
        halop = 15 ;
        lev = 92 ;
        levp = 93 ;

**Reply by evankalina** [reply]
@lliu1220 Thanks! I'm trying to figure out how to generate the gfs_bndy.tile7 file with these dimensions. What does the grid look like in your experiment? Is it the regional HAFS grid? If not, can you point me to your model output?

**Reply by lliu1220** [reply]
@evankalina
thank you! These model output are at: 
/work/noaa/hwrf/scrub/lliu/HAFS_ghrsst/2019082900/00L/forecast/INPUT/gfs_bndy.tile7.0*.nc
they are at on = 2888 ;
lat = 1920 ;

My question is how to output the lower resolution:
lon = 728 ;
lat = 600 ;

I do not think this is related to regional ocean resolution, I checked both low and high gfs bndy experiments, regional resolution has alwasy been:
dimensions:
        time = UNLIMITED ; // (10 currently)
        Latitude = 633 ;
        Longitude = 1135 ;
variables:

**Reply by evankalina** [reply]
Hi @lliu1220, thanks! But do you have model output from a run in which the gfs_bndy.tile7.0*.nc has the desired lower resolution:
lon = 728 ;
lat = 600 ;

If I could see the settings used in chgres_cube to generate the lower-resolution gfs_bndy file, I could implement them in the HAFS workflow.

**Comment by BinLiu-NOAA** [comment, marked as answer]
@evankalina and @lliu1220, if your test is based on @uturuncoglu's ufs-weather-model level regression test, then most likely, this is the HAFS application level configuration I used to generate the atmospheric model needed grid, IC/BC. 
Orion: 
vi /work/noaa/hwrf/save/bliu/hafs_develop_202102/rocoto/cronjob_hafs_rt_test.sh
 ${PYTHON3} ./run_hafs.py -t ${dev} 2019082900 00L HISTORY \
     config.EXPT=${EXPT} config.SUBEXPT=${EXPT}_rt_regional_C192_1n4_static_new \
     config.NHRS=126 ${scrubopt} \
     **../parm/hafsC192_1n4_L91_AL.conf** \
     ../parm/hafs_hycom.conf

 ${PYTHON3} ./run_hafs.py -t ${dev} 2019082900 00L HISTORY \
     config.EXPT=${EXPT} config.SUBEXPT=${EXPT}_rt_regional_C192_1n4_static_atmonly_new \
     config.NHRS=126 ${scrubopt} \
     ../parm/hafsC192_1n4_L91_AL.conf

**Reply by evankalina** [reply]
@BinLiu-NOAA Thanks! I should have thought to look at the existing conf files. Hopefully this will work for Ling.

**Reply by lliu1220** [reply]
@BinLiu-NOAA 
I am using your crontab, I changed everything under hwrf and hurricane, but my submission keep sending me to aoml_hafs1, any idea why? Thanks

/work/noaa/hwrf/save/lliu/hafs_develop_202102/rocoto/cronjob_hafs_rt_test.sh

#! /bin/sh
#SBATCH --job-name=hafs_launch_00L_2019082900
#SBATCH -o /work/noaa/hwrf/scrub/lliu/hafs_develop_202102_rt_regional_C192_1n4_static_atmonly_new/2019082900/00L/hafs_launch.log
#SBATCH --account=aoml-hafs1
#SBATCH --qos=batch
#SBATCH --ntasks=1
#SBATCH -t 00:15:00
#SBATCH --mem=1024
#SBATCH --partition=orion
#SBATCH --comment=2565c8bcb92f2b522a16e74b4c244931
export TOTAL_TASKS='1'
export envir='prod'

**Reply by evankalina** [reply]
@lliu1220 The account key is specified in parm/system.conf. Does it say aoml-hafs1 in there?

**Reply by lliu1220** [reply]
@evankalina thank you!
no, it doesn't:
disk_project=hwrf
## Project hpss tape area
tape_project=emc-hwrf
## Theia CPU account name for submitting jobs to the batch system.
cpu_account=hurricane

**Reply by lliu1220** [reply]
@BinLiu-NOAA @evankalina 

I made copy of /work/noaa/hwrf/save/bliu/hafs_develop_202102
and changed under rocoto and parm directories.

I wonder if this is the right way of making this setup work,
Can you share with us a README on how to make this configuration work?

Thanks.

**Reply by BinLiu-NOAA** [reply]
@lliu1220, it's better to checkout the latest HAFS develop branch or the feature/hafs_sync_202107 branch. But, anyway, if you copied over my directory, you need to change this 'ENTITY ACCOUNT         "aoml-hafs1"' into 'ENTITY ACCOUNT         "&CPU_ACCOUNT;"' in rocoto/sites/orion.ent. And then follow @evankalina's instruction above to specify the cpu_account properly (default is hurricane on Orion).

BTW, if you want to run the HAFS CDEPS workflow, you probably want to checkout the feature/hafs_couplehycom_cdeps branch. Also @evankalina it will be nice if you sync this feature/hafs_couplehycom_cdeps with the latest HAFS develop branch when you get a chance.

Thanks!

**Reply by lliu1220** [reply]
@BinLiu-NOAA 
We already did. I was able to runthrough cdeps for 24 hour forecast with your version of lateral boundary condition.
The reason I'm running this develop branch is to create same resolution of 5 day forecast of lateral boundary condition
according to @evankalina and Ufuk this is the right thing to do.
But it still is not working.
Btw, I checked under your rocoto: hafs_workflow.xml.in is where to specify ENTITY ACCOUNT, yes?
  <!ENTITY CPU_ACCOUNT "@[CPU_ACCOUNT]">
I did not find any 'aoml' keywords, and it seemed already specified as CPU_ACCOUNT.

Thanks!

**Reply by BinLiu-NOAA** [reply]
@lliu1220 it's in this file: **rocoto/sites/orion.ent**

**Reply by evankalina** [reply]
@lliu1220 and @BinLiu-NOAA As of this morning, the feature/hafs_couplehycom_cdeps branch is synced with the latest HAFS develop branch, and we have issued a PR to pull it into develop.

**Comment by lliu1220** [comment]
@BinLiu-NOAA 

I'm currently using 20200825 hurricane data under your this directory-
/work/noaa/hwrf/scrub/bliu/hafsv0p2a_phase3_final/com/

I wonder how likely to produce these hurricane data for 20200720, Hurricane Isaias Initial condition, can you please instruct?
Thank you.

**Reply by evankalina** [reply]
Hi @lliu1220,

You should be able to run the HAFS workflow for any recent case that you are interested in, including Isaias.

Would you please take a look at [the documentation](https://docs.google.com/document/d/1_UsLo96LAxCZjXHZe6qrr2WsR5W8PC3CGJxHod9qWKE/edit) for running HAFS DOCN using the workflow? If you run the workflow through the atm_ic and atm_lbc tasks, you should have the necessary initial and lateral boundary conditions for running the rest of this case offline. Remember to include the "../parm/hafsC192_1n4_L91_AL.conf" file in your cronjob if you want to obtain C192 GFS ICs.

**Reply by lliu1220** [reply]
@uturuncoglu @evankalina @BinLiu-NOAA 

I ran through hafs coupled hycom branch , and I was able to obtain hurricane **trak** analysis.
For instance @BinLiu-NOAA 's config: /work/noaa/hwrf/save/bliu/hafs_develop_202102/rocoto/cronjob_hafs_rt_test.sh

However, I do not see any hurricane trak output from HAFS_CDEPS_DOCN. I was also able to run this config, and pass regression test for DOCN.

My question to @uturuncoglu @evankalina 
Does DOCN have hurricnae track forecast or not?

Thanks.
Ling

**Reply by evankalina** [reply]
@lliu1220 Yes, CDEPS DOCN runs generate ATCF track files if you are using the workflow from the feature/hafs_couplehycom_cdeps branch. The ATCF output gets created when the product job runs. It will get copied to a directory similar to this one:
/work/noaa/dtc-hwrf/ekalina/noscrub/hafstrak

For example, here is my ATCF track file from a DOCN run with OISST data:
/work/noaa/dtc-hwrf/ekalina/noscrub/hafstrak/hafs_couplehycom_cdeps-20210813_oisst/natl00l.2019082900.trak.hafs.atcfunix.all

If you don't see this file, would you please check to see if the product job ran in the workflow?

**Reply by lliu1220** [reply]
Thanks! @evankalina , 
I had look at your oisst.conf , it seems you have global mesh, does this version read in netCDF regional file input ?
Also can you give permisison to /work/noaa/dtc-hwrf/ekalina/noscrub/hafstrak

/work/noaa/dtc-hwrf/ekalina/hafs_couplehycom_cdeps-20210813
I want to have a look at your config.

Thank you!

[config]
docn_source=OISST

[forecast]
docn_mesh_nx_global=1440
docn_mesh_ny_global=720
mesh_ocn_in={FIXmeshes}/docn_oisst_mesh.nc ;; premade mesh to use if make_mesh_ocn=no

**Reply by evankalina** [reply]
@lliu1220 The permissions should be correct now. The downloaded OISST dataset has global coverage, but the data are being used in a regional HAFS DOCN experiment.

**Reply by lliu1220** [reply]
@evankalina 
Thank you for your help and issuing permission. Now that I'm able to run this branch.
I see there is a script under /work/noaa/hwrf//save/lliu/hafs_couplehycom_cdeps-20210813/ush/hafs_oisst_download.py
that download the oisst data , my question is, are these data like the HYCOM data used as initial condition in previous branch I tested : 
/work/noaa/marine/lliu/HAFS_couplling/cpld_regional_natl_c192_docn_mom6/INPUT_DOCN/hycom_201908_surf_nolev.nc
If yes, then how to enable this feature (HYCOM netcdf initial condition) in this branch? Can you please advise?
Appreciate your time, -Ling

**Reply by lliu1220** [reply]
@uturuncoglu @BinLiu-NOAA 
Also added Ufuk and Bin, Please let me know if you have an answer for this question, waiting--

there is a script under /work/noaa/hwrf//save/lliu/hafs_couplehycom_cdeps-20210813/ush/hafs_oisst_download.py
that download the oisst data , my question is, are these data like the HYCOM data used as initial condition in previous branch I tested :
/work/noaa/marine/lliu/HAFS_couplling/cpld_regional_natl_c192_docn_mom6/INPUT_DOCN/hycom_201908_surf_nolev.nc
If yes, then how to enable this feature (HYCOM netcdf initial condition) in this branch? Can you please advise?
Appreciate your time, -Ling

**Reply by evankalina** [reply]
@lliu1220 I am not sure if I fully understand your question. OISST is an observations-based analysis of SST that uses data from many platforms (see [link](https://www.ncdc.noaa.gov/oisst/optimum-interpolation-sea-surface-temperature-oisst-v21)), while HYCOM data are an SST analysis (at t=0h) or forecast (at t>0h) generated by that modeling system. So they are different. If you are asking if it is possible to prepare DOCN runs that use the output from HYCOM, it can likely be done, but it isn't supported by the workflow itself. You would need to follow the steps in Appendix B ("Considerations for adding a new dataset") in the HAFS-CDEPS documentation.

**Comment by lliu1220** [comment]
@evankalina 
Can you take a look at this config, which regression test was performed?
It uses HYCOM as initial condition, all I'm asking is whether your branch hafs_couplehycom_cdeps-20210813 can do the same:
/work/noaa/marine/lliu/HAFS_couplling/cpld_regional_natl_c192_docn_mom6/INPUT_DOCN/

So I took it as no, for now, correct me if i'm wrong.
Thanks

**Reply by evankalina** [reply]
@lliu1220 OK, thanks for clarifying. Would you please open read permissions?

**Reply by evankalina** [reply]
@lliu1220 Since you've already successfully run this configuration using HYCOM data, I see no reason why you couldn't use the code from the feature/hafs_couplehycom_cdeps branch to do it. The branch is up to date with all of the recent changes to ufs-weather-model.

**Reply by lliu1220** [reply]
@evankalina 
Unfortuantely that version does not have hurricane track forecast. 
that was why I'm running your branch, it does have hurrcinae trak forecast, but not netcdf initial condition.
is there anything can be done to make at least one version work for hurrcane track forecast using NetCDF initial condition?
Thanks.

**Reply by evankalina** [reply]
Closing the loop on this discussion, since it was addressed offline. In short, the answer to Ling's question is:

"The hurricane track forecast is produced by the "product" task in the HAFS workflow, which requires inputs from the "atm_post" task. If you want the hurricane track forecast, you will need to run the full workflow, rather than the offline regression test."

Ling wanted to create a data ocean using output from MOM6, which isn't officially supported in the workflow. But since she already prepared the MOM6 input data and the mesh file, it was straightforward to adapt the workflow to her needs:

"Basically, you can pretend that you are using OISST data, and the workflow won't know the difference. Here are the steps:

1. Modify scripts/exhafs_docn_prep.sh slightly to bypass the oisst_prep script, since your data are already prepared. The only thing that the workflow needs to do is copy your data to the working directory. Assuming your input dataset is called hycom_201908_surf_nolev.nc:
 if [[ "$docn_source" == OISST ]] ; then   
 -$APRUNS "$USHhafs/cdeps_utils/hafs_oisst_prep.sh" "$docn_input_path"
 +$USHhafs/produtil_deliver.py -c "$DOCNdir/hycom_201908_surf_nolev.nc" "$docn_input_path/DOCN_input_00000.nc"
 elif [[ "${docn_source}" == RTOFS ]] ; then
     $APRUNS "$USHhafs/cdeps_utils/hafs_rtofs_prep.sh" "$docn_input_path"
 elif [[ "${docn_source}" == GHRSST ]] ; then

2. Modify parm/cdeps/docn_oisst.streams, because the SST variable in your dataset is called temp, not sst:
-stream_data_variables01:   "sst So_t"
+stream_data_variables01:   "temp So_t"

3. Edit rocoto/cronjob_hafs_cdeps.sh to run the DOCN OISST experiment. You can comment out all of the other experiments. The run command inside the script should include the path to your input data directory, and the path and name of your mesh file:
${PYTHON3} ./run_hafs.py -t ${dev} 2019082318 00L HISTORY config.EXPT=${EXPT} \
    config.SUBEXPT=${EXPT}_hat10 \
    forecast.output_history=.true. \
    ../parm/hafs_regional_static.conf \
    ../parm/hafs_docn.conf ../parm/hafs_docn_oisst.conf \
    config.NHRS=120 \
    dir.DOCNdir=/work/noaa/dtc-hwrf/ekalina/DOCN_DATA/from_ling \
    forecast.mesh_ocn_in=/work/noaa/dtc-hwrf/ekalina/DOCN_DATA/from_ling/hat10_210129_ESMFmesh_py.nc"

**Reply by lliu1220** [reply]
@evankalina 
Thank you, Evan , for your diligent help to make this work possible!

**Comment by lliu1220** [comment]
@evankalina 
I'm so sorry! should be correct now
